src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs

```diff
@@ -128,203 +129,194 @@ public unsafe static uint[] Remainder(uint[] left, uint[] right)
+            Debug.Assert(rightLength >= 0);
+            Debug.Assert(leftLength >= rightLength);
+            Debug.Assert(q <= 0xFFFFFFFF);
 
-            if (shift > 0)
-            {
-                int backShift = 32 - shift;
-                for (int i = 0; i < valueLength - 1; i++)
-                {
-                    target[i] = (value[i] >> shift)
-                        | (value[i + 1] << backShift);
-                }
-                target[valueLength - 1] =
-                    (value[valueLength - 1] >> shift);
-            }
-            else
+            // Combines a subtract and a multiply operation, which is naturally
+            // more efficient than multiplying and then subtracting...
+
+            int i = 0;
+            ulong carry = 0L;
+
+            for (; i < rightLength; i++)
             {
-                for (int i = 0; i < valueLength; i++)
-                    target[i] = value[i];
+                carry += right[i] * q;
+                uint digit = (uint)carry;
+                carry = carry >> 32;
+                if (left[i] < digit)
+                    ++carry;
+                left[i] -= digit;
             }
+
+            return (uint)carry;
         }
 
-        [SecuritySafeCritical]
-        private unsafe static void MultiplyDivisor(uint* left, int leftLength,
-                                                   ulong right, uint* bits)
+        private static bool CompareDivisor(uint divHi, uint divLo,
+                                           ulong valHi, uint valLo,
+                                           ulong q)
         {
-            Debug.Assert(leftLength >= 1);
-            Debug.Assert(right <= 0xFFFFFFFF);
+            Debug.Assert(q <= 0xFFFFFFFF);
 
-            ulong carry = 0UL;
-            for (int i = 0; i < leftLength; i++)
-            {
-                ulong digits = left[i] * right + carry;
-                bits[i] = (uint)digits;
-                carry = digits >> 32;
-            }
-            bits[leftLength] = (uint)carry;
+            // We multiply the two most significant limbs of the divisor
+            // with the current guess for the quotient. If those are bigger
+            // than the three most significant limbs of the current dividend
+            // we return true, which means the current guess is still too big.
+
+            ulong chkHi = divHi * q;
+            ulong chkLo = divLo * q;
+
+            chkHi = chkHi + (chkLo >> 32);
+            chkLo = chkLo & 0xFFFFFFFF;
+
+            if (chkHi < valHi)
+                return false;
+            if (chkHi > valHi)
+                return true;
+
+            if (chkLo < valLo)
+                return false;
+            if (chkLo > valLo)
+                return true;
+
+            return false;
+        }
+
+        private static uint[] CreateCopy(uint[] value)
+        {
+            Debug.Assert(value != null);
+
+            var bits = new uint[value.Length];
```

Comment:
Will it be common for ```value.Length == 0``` ?  If no, nevermind.  If yes, consider special-casing it to return ```Array.Empty<uint>()```.